---
title: "Transcriptomics and the analysis of RNA-Seq data"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Import countData and colData
```{r eval=FALSE}
BiocManager::install("AnnotationDbi")
BiocManager::install("org.Hs.eg.db")
BiocManager::install("EnhancedVolcano")
```

```{r}
library(DESeq2)
counts <- read.csv("data/airway_scaledcounts.csv", stringsAsFactors = FALSE)
metadata <-  read.csv("data/airway_metadata.csv", stringsAsFactors = FALSE)
table(metadata$dex)
all(colnames(counts)[-1] == metadata$id) # See if all TRUE
```
```{r eval=FALSE}
head(counts)
nrow(counts) #number of genes: 38694
head(metadata)
```

## Toy differential gene expression
```{r}
control <- metadata[metadata[,"dex"]=="control",]
control.mean <- rowSums( counts[ ,control$id] )/length(control$id)
names(control.mean) <- counts$ensgene
treated <- metadata[metadata[,"dex"]=="treated",]
treated.mean <- rowSums(counts[,treated$id])/length(treated$id)
names(treated.mean) <- counts$ensgene
meancounts <- data.frame(control.mean, treated.mean)
plot(control.mean, treated.mean, xlab="Control", ylab="Treated")
plot(control.mean, treated.mean,log = "xy", xlab="Control", ylab="Treated")
```

```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```

```{r eval=FALSE}
#Test
x <- c(3,4,0,5,0)
x == 0
which(x == 0)
y <- data.frame(happy=c(5,6,0,0), sad=c(0,5,5,0))
y == 0
which(y==0, arr.ind = T) #give the row and col position of 0
ind <- which(y==0, arr.ind = T)[,1]#focus the row only that have 0 entries
ind2 <- unique(ind)#remove duplicated rows
y[-ind2,]#select non-zero entires
```

## Exclude zero entries
```{r}
zero.vals <- which(meancounts[,1:2]==0, arr.ind=TRUE)

to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
nrow(mycounts)
```

## Identify up/down regulated entries
```{r}
up.ind <- mycounts$log2fc > 2
#length(which(up.ind == TRUE)) 
sum(up.ind)
down.ind <- mycounts$log2fc < (-2)
#length(which(down.ind == TRUE))
sum(down.ind)
```

## Adding annotation data
```{r}
library(AnnotationDbi)
library(org.Hs.eg.db)
anno <- read.csv("data/annotables_grch38.csv")
head(anno)
```
```{r}
mycounts$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(mycounts), # Our genenames
                     keytype="ENSEMBL",        # The format of our genenames
                     column="SYMBOL",          # The new format we want to add
                     multiVals="first")
head(mycounts)
```

## DESeq2 analysis
```{r}
library(DESeq2)
citation("DESeq2")
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex, 
                              tidy=TRUE)
dds <- DESeq(dds)
res <- results(dds)
res
summary(res)
resOrdered <- res[order(res$pvalue),]
res05 <- results(dds, alpha=0.05)
summary(res05)
resSig05 <- subset(as.data.frame(res), padj < 0.05)
nrow(resSig05)
ord <- order( resSig05$padj )
#View(res01[ord,])
head(resSig05[ord,])
write.csv(resSig05[ord,], "signif05_results.csv")
```

## Data Visualization
```{r}
plot( res$log2FoldChange,  -log(res$padj), 
      xlab="Log2(FoldChange)",
      ylab="-Log(P-value)")
plot( res$log2FoldChange,  -log(res$padj), 
 ylab="-Log(P-value)", xlab="Log2(FoldChange)")

# Add some cut-off lines
abline(v=c(-2,2), col="darkgray", lty=2)
abline(h=-log(0.05), col="darkgray", lty=2)
```
```{r}
# Add colors
# Setup our custom point color vector 
mycols <- rep("gray", nrow(res))
mycols[ abs(res$log2FoldChange) > 2 ]  <- "red" 

inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

# Volcano plot with custom colors 
plot( res$log2FoldChange,  -log(res$padj), 
 col=mycols, ylab="-Log(P-value)", xlab="Log2(FoldChange)" )

# Cut-off lines
abline(v=c(-2,2), col="gray", lty=2)
abline(h=-log(0.1), col="gray", lty=2)
```

```{r}
# Add annotation
x <- as.data.frame(res)
x$symbol <- mapIds(org.Hs.eg.db, 
                   keys=row.names(x),
                   keytype="ENSEMBL",
                   column="SYMBOL",
                   multiVals="first")
library(EnhancedVolcano)
EnhancedVolcano(x,
    lab = x$symbol,
    x = 'log2FoldChange',
    y = 'pvalue')
```

